addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-synchronizer.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"flag"
+	"github.com/aws/aws-sdk-go/aws"
+	"log"
+	"path/filepath"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+)
+
+func main() {
+	defaultS3Mounts, region, profile, destinationBase, concurrency, debug := readConfigFromArgs()
+	sess := makeSession(profile, region)
+
+	mainImpl(sess, debug, concurrency, defaultS3Mounts, destinationBase)
+}
+
+func mainImpl(sess *session.Session, debug bool, concurrency int, defaultS3Mounts string, destinationBase string) error {
+	// Use a map to emulate a set to keep track of existing mounts
+	currentMounts := make(map[string]struct{}, 0)
+	mountsCh := make(chan *mountConfiguration, 50)
+
+	// Create wait group to keep track of go routines being spawned
+	// so the main go routine can wait for them to completes
+	var wg sync.WaitGroup
+
+	// In another thread, get the next mount configuration from the buffered channel
+	// and download the files. If the share is marked as writeable then start the
+	// file watchers in another thread (because the setup function won't return)
+	go func() {
+		for {
+			mountConfig := <-mountsCh
+			if debug == true {
+				log.Printf("Received mount configuration from channel: %+v\n", mountConfig)
+			}
+			downloadFiles(sess, mountConfig, concurrency, debug)
+			if mountConfig.writeable {
+				go func() {
+					err := setupUploadWatcher(sess, mountConfig, debug)
+					if err == nil {
+						wg.Add(1) // Increment wait group counter everytime we spawn upload watcher
+					} else {
+						log.Printf("Error setting up file watcher: " + err.Error())
+					}
+				}()
+			}
```

Comment:
I missed seeing this before somehow. Do we need ship upload-watcher and even the initialization here if that is unfinished?

Reply:
Done. Will add this along with unit tests when adding Read/Write functionality.